stdlib/public/core/Result.swift

```diff
@@ -0,0 +1,240 @@
+  ///   instance.
+  /// - Returns: A new `Result` instance, either from the transform or from
+  ///   the previous success value.
+  public func flatMapError<NewError>(
+    _ transform: (Error) -> Result<Value, NewError>
+  ) -> Result<Value, NewError> {
+    switch self {
+    case let .success(value):
+      return .success(value)
+    case let .failure(error):
+      return transform(error)
+    }
+  }
+  
+  /// Evaluates the given transform closures to create a single output value.
+  ///
+  /// - Parameters:
+  ///   - onSuccess: A closure that transforms the success value.
+  ///   - onFailure: A closure that transforms the error value.
+  /// - Returns: A single `Output` value.
+  public func fold<Output>(
+    onSuccess: (Value) -> Output,
+    onFailure: (Error) -> Output
+  ) -> Output {
+    switch self {
+    case let .success(value):
+      return onSuccess(value)
+    case let .failure(error):
+      return onFailure(error)
+    }
+  }
+}
+
+extension Result where Error : Swift.Error {
+  /// Unwraps the `Result` into a throwing expression.
+  ///
+  /// - Returns: The success value, if the instance is a success.
+  /// - Throws:  The error value, if the instance is a failure.
+  public func unwrapped() throws -> Value {
+    switch self {
+    case let .success(value):
+      return value
+    case let .failure(error):
+      throw error
+    }
+  }
+}
+
+extension Result where Error == Swift.Error {
+  /// Create an instance by capturing the output of a throwing closure.
+  ///
+  /// - Parameter throwing: A throwing closure to evaluate.
+  @_transparent
+  public init(_ throwing: () throws -> Value) {
+    do {
+      let value = try throwing()
+      self = .success(value)
+    } catch {
+      self = .failure(error)
+    }
+  }
+  
+  /// Unwraps the `Result` into a throwing expression.
+  ///
+  /// - Returns: The success value, if the instance is a success.
+  /// - Throws:  The error value, if the instance is a failure.
+  public func unwrapped() throws -> Value {
+    switch self {
+    case let .success(value):
+      return value
+    case let .failure(error):
+      throw error
+    }
+  }
+  
+  /// Evaluates the given transform closure when this `Result` instance is
+  /// `.success`, passing the value as a parameter and flattening the result.
+  ///
+  /// - Parameter transform: A closure that takes the successful value of the
+  ///   instance.
+  /// - Returns: A new `Result` instance, either from the transform or from
+  ///   the previous error value.
+  public func flatMap<NewValue>(
+    _ transform: (Value) throws -> NewValue
+  ) -> Result<NewValue, Error> {
+    switch self {
+    case let .success(value):
+      do {
+        return .success(try transform(value))
+      } catch {
+        return .failure(error)
+      }
+    case let .failure(error):
+      return .failure(error)
+    }
+  }
+}
+
+extension Result : Equatable where Value : Equatable, Error : Equatable { }
```

Comment:
Would it make sense to provide `~=` (switch-matching) operators for `Value: Equatable` and `Error` patterns? (Hmm, technically that should be part of the proposal if it's included...)